src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_ReturnKey.cs

```diff
@@ -64,7 +65,8 @@ private void CommitOnEnter(out bool sendThrough, out bool committed)
             // to the editor.  In single-line debugger windows (Watch, etc), however, we don't
             // want to send the enter though, because those windows don't support displaying
             // more than one line of text.
-            sendThrough = !model.TriggerInfo.IsDebugger || model.TriggerInfo.IsImmediateWindow;
+            sendThrough = !model.Trigger.CustomTags.Contains(WellKnownCompletionTriggerTags.Debugger)
```

Comment:
@jasonmalinowski 

Those aren't terrible ideas at all! :smile:

>I think you already had some sort of "context" object that was beginning to congeal. Perhaps a type that is the "context" that contains the trigger, various "environment" flags (notably this debugger flag and the interactive flag), and maybe some other pieces of information could become the replacement.

This would need to be a *yet another* type than the "context" object that is passed by the completion service to ```CompletionListProvider.ProduceCompletionListAsync```. I'd prefer not to introduce more of these. However, your point is taken. I was modeling CustomTags off of what we did with diagnostics, but perhaps that's not the best approach here.

>Is the debugger flag only being used to control whether we list completion items before their declaration point? And do completion providers get passed an OptionSet from the originating workspace (or can get at it)? If both are "yes", then perhaps we should instead have there be a not-UI-exposed completion option that simply specifies that debugger behavior, and the debugger workspace we create just sets that option explicitly. I could imagine people might actually like that behavior during real typing as well. It's not rare that you might write some code and realize you need to use a local declared farther down. Rather than me breaking context, moving the local up, and then writing the code again, that it's perhaps OK to actually complete it anyways. It's I am not saying we should expose it in a UI (at least not until customers come with pitchforks asking for it :smile:), but if they did come along and ask for that we'd implement that with an option. Given that also might fix this problem too, I can't help but notice the coincidence.

This is the most appealing idea. Rather than having flags like IsDebugger and IsImmediateWindow that control random behaviors, capturing those behaviors in options is a good idea. I'll try this one out.

>Rather than having some additional way to shuttle info, this could just be checking the workspace kind. It's still comparing magic strings, but at least without the baggage of extra types.

Workspace kind is a pretty big hammer that I'd like to avoid. Of course, you probably listed it last because you'd like me to avoid it too. :smile: